cumulus/client/relay-chain-rpc-interface/src/reconnecting_ws_client.rs

```diff
@@ -460,4 +497,48 @@ mod test {
 			url_to_string_with_port(url)
 		);
 	}
+
+	#[tokio::test]
+	// Testing the retry logic at full means increasing CI with half a minute according
+	// to the current logic, so lets test it best effort.
+	async fn client_manager_retry_logic() {
+		let port = portpicker::pick_unused_port().unwrap();
+		let server = jsonrpsee::server::Server::builder()
+			.build(format!("0.0.0.0:{}", port))
+			.await
+			.unwrap();
+
+		// Wait three seconds while attempting connection.
+		let conn_res = tokio::spawn(async move {
+			tokio::time::timeout(
+				Duration::from_secs(3),
```

Comment:
Using sleep in tests has high potential to make it flaky. I don't understand why we wait here 3 seconds, only to expect the obvious, there is no server up to connect to. What can happen is that the server might actually start if CI is overloaded and this will fail.

For testing the logic here I think there are other ways to do it. I mean you just want to see it trying multiple servers in the list and increasing the back off. This stuff is not actually tested here at all.

Reply:
Haven't seen this. It came a bit earlier than this: https://github.com/paritytech/polkadot-sdk/pull/5515/commits/71e44b47cc4514f659a3650b8125ebc7143f58c8. I agree with your points.

> I don't understand why we wait here 3 seconds, only to expect the obvious, there is no server up to connect to.

The idea of the wait is that the `ClientManager::new` call doesn't exit immediately as it should've (because of the connection refused error), but it exercises the retry logic hoping the server would be up. The error we're asserting against is the `Elapsed` error which is triggered because the `ClientManager::new` call timed out - which means retry logic worked behind the scenes.

I think though we can simplify the test by starting the server in a tokio task with a 10 second delay (10 because it is not too big/small for our usecase), and then create a client on the "main" task which exercises the retry logic until the server gets up and returns with an `Ok` result.

> I mean you just want to see it trying multiple servers in the list and increasing the back off. This stuff is not actually tested here at all.

I think testing the backoff involves some other timers with timeouts given the logic doesn't produce testable side effects, which can add more sources of flakyness. I am open to suggestions if you see it differently. As a matter of fact I wasn't interested in testing the backoff, but the fact that the collators' relay chain interface tries a few more times to connect to the external RPCs at collator startup. I heard there weren't complaints about this in practice meaning the collators usually find available RPC nodes to connect to. This is more relevant to local testing on slow/overloaded machines, so testing anything other than a slight reliability improvement brings diminishing returns.

Reply:
Pushed a simpler version of the test here: [ee68d04](https://github.com/paritytech/polkadot-sdk/pull/5515/commits/ee68d040653b801897af8e646c19508a8015511d). Let me know if this changes anything!